### ISSUE TITLE
Motivation?

### ISSUE THREAD

**Issue by 3noch** [composewell/streamly#2]
Looks very neat! You mention in the README that this library is inspired by `transient`. What caused you to write this library instead of using `transient`, for example? What other motivations led to this work? (I'd recommend adding the answers to these questions directly to the README.)

**Comment by harendra-kumar** [comment]
I was about to write about this in the tutorial (which is incomplete and incorrect at this point), but you asked already. The original motivation was to create a transformer that can be used anywhere, transient if a monolithic monad. But then I wanted it to be lazy and streaming as well rather than strict so that I could stream and adapt the threads based on how fast the consumer is running. The result is that it became a first class streaming library with first class concurrency and when we add these two it becomes first class FRP as well. Neat! Though I have not added convenience APIs for FRP yet, it can still do quite a bit with just the core.

I will add it to the "comparison with other libraries" section or a little bit in the readme as well. I will associate a gitter channel as well for discussions and suggestions.

**Comment by harendra-kumar** [comment]
Hey @3noch, I updated the README with examples and the tutorial in haddock is now up to date (`stack haddock --open` should open the docs in your browser). The tutorial has a "comparison with other packages" section. I also created a gitter channel for questions/discussions -  https://gitter.im/composewell/streamly .

**Comment by 3noch** [comment]
Fantastic. Feel free to consider this issue resolved!

**Comment by 3noch** [comment]
I might also recommend that you post to Reddit.

**Comment by harendra-kumar** [comment]
Let me upload it to hackage first, so that people can conveniently look at the docs and try it.

**Comment by harendra-kumar** [comment]
@jberryman thanks for reporting your use case here, I will take a look and see what best we can do in this case.

**Comment by harendra-kumar** [comment]
@jberryman I took a look at your case. I think streamly can do what you wanted to do, idiomatically. You can just say something like `coparallely $ cycle1 $ once ioaction`(the API names used here are not in the released code, will be released as 0.2.0 soon). Except that #66 is important to fix before you can use it, to put a bound on how many results you want buffered; we may have a combinator for that so that the programmer can tune it. Also, perhaps you want something like execute ahead but yield in order for which I have #53. The current types `Parallel` and `Coparallel` yield the results as they arrive and not in the order in which you triggered them.

If that is something that you were looking for then I will close this issue.

**Comment by harendra-kumar** [comment]
@jberryman I introduced an `Ahead` stream type that produces results in the traversal order but can execute multiple actions concurrently. So you can write something like this:

```haskell
import Streamly
import qualified Streamly.Prelude as S
import Control.Concurrent
import Data.Semigroup

main = runStream $ do
    x <- aheadly $ cycle1 (delay 10)
    S.once $ print x
    delay 1

    where delay n = S.once $ do
             threadDelay (n * 1000000)
```

Here we are executing a task with long delay (10 sec) using `aheadly` combinator, it will keep the stream ready for consumption by the rest of the loop which consumes it 1 result per second. You can use `asyncly` instead of `aheadly` if you don't care about the order of the results produced by the concurrent actions.
